src/autoprops.rs

```diff
@@ -0,0 +1,261 @@
+                     (#[autoprops] must be placed *before* #[function_component])",
+                )
+            })?;
+
+        for input in &sig.inputs {
+            if let syn::FnArg::Typed(syn::PatType { pat, .. }) = input {
+                if let syn::Pat::Wild(wild) = pat.as_ref() {
+                    return Err(syn::Error::new_spanned(
+                        wild,
+                        "cannot use `_` as field name",
+                    ));
+                }
+            }
+        }
+
+        let properties_name = syn::Ident::new(
+            &format!("{}Props", component_name),
+            proc_macro2::Span::call_site(),
+        );
+
+        Ok(Self {
+            properties_name,
+            item_fn,
+        })
+    }
+}
+
+impl Autoprops {
+    pub fn apply_args(&mut self, args: AutopropsArgs) {
+        if let Some(name) = args.properties_name {
+            self.properties_name = name;
+        }
+    }
+
+    fn print_function_component(&self) -> proc_macro2::TokenStream {
+        let properties_name = &self.properties_name;
+        let syn::ItemFn {
+            attrs,
+            vis,
+            sig,
+            block,
+        } = &self.item_fn;
+
+        let fn_name = &sig.ident;
+        let (_impl_generics, type_generics, where_clause) = sig.generics.split_for_impl();
+        let inputs = if self.needs_a_properties_struct() {
+            // NOTE: function components currently don't accept receivers, we're just passing the
+            //       information to the next macro to fail and give its own error message
+            let receivers = sig
+                .inputs
+                .iter()
+                .filter_map(|arg| match arg {
+                    syn::FnArg::Receiver(receiver) => Some(receiver),
+                    _ => None,
+                })
+                .collect::<Vec<_>>();
+            let args = sig
+                .inputs
+                .iter()
+                .filter_map(|arg| match arg {
+                    syn::FnArg::Typed(syn::PatType { pat, .. }) => Some(quote! { #pat, }),
+                    _ => None,
+                })
+                .collect::<Vec<_>>();
+            let phantom = sig.generics.type_params().next().is_some().then(|| {
+                quote! {
+                    phantom: _,
+                }
+            });
+            quote! {
+                #(#receivers,)* #properties_name {
+                    #(#args)*
+                    #phantom
+                }: &#properties_name #type_generics
+            }
+        } else {
+            quote! {}
+        };
+        let clones = sig
+            .inputs
+            .iter()
+            .filter_map(|arg| match arg {
+                syn::FnArg::Typed(syn::PatType { pat, ty, .. })
+                    if !matches!(**ty, syn::Type::Reference(_)) =>
+                {
+                    Some(quote! { let #pat = ::std::clone::Clone::clone(#pat); })
```

Comment:
I don't think the clone should be implicit here. If the clone is to be implicit, it should call `ImpicitClone` so when anything that doesn't implement `ImpicitClone` is attempted to be implicitly cloned, there is a compiler error

Reply:
I can't figure out how to call it :persevere: halp please

```
fn main() {
    let x = 42_u32;
    <::implicit_clone::ImplicitClone as Clone>::clone(&x);
}
```

```
error[E0038]: the trait `ImplicitClone` cannot be made into an object
 --> src/main.rs:3:55
  |
3 |     <::implicit_clone::ImplicitClone as Clone>::clone(&x);
  |                                                       ^^ `ImplicitClone` cannot be made into an object
  |
  = note: the trait cannot be made into an object because it requires `Self: Sized`
  = note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
  = note: required for the cast from `&u32` to `&dyn ImplicitClone`

error[E0277]: the size for values of type `dyn ImplicitClone` cannot be known at compilation time
   --> src/main.rs:3:55
    |
3   |     <::implicit_clone::ImplicitClone as Clone>::clone(&x);
    |     ------------------------------------------------- ^^ doesn't have a size known at compile-time
    |     |
    |     required by a bound introduced by this call
    |
    = help: the trait `Sized` is not implemented for `dyn ImplicitClone`
note: required by a bound in `clone`
   --> /home/cecile/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/clone.rs:147:18
    |
147 | pub trait Clone: Sized {
    |                  ^^^^^ required by this bound in `Clone::clone`
...
160 |     fn clone(&self) -> Self;
    |        ----- required by a bound in this associated function

error[E0038]: the trait `ImplicitClone` cannot be made into an object
 --> src/main.rs:3:6
  |
3 |     <::implicit_clone::ImplicitClone as Clone>::clone(&x);
  |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `ImplicitClone` cannot be made into an object
  |
  = note: the trait cannot be made into an object because it requires `Self: Sized`
  = note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>

error[E0038]: the trait `ImplicitClone` cannot be made into an object
 --> src/main.rs:3:5
  |
3 |     <::implicit_clone::ImplicitClone as Clone>::clone(&x);
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `ImplicitClone` cannot be made into an object
  |
  = note: the trait cannot be made into an object because it requires `Self: Sized`
  = note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>

error[E0746]: return type cannot have an unboxed trait object
 --> src/main.rs:3:5
  |
3 |     <::implicit_clone::ImplicitClone as Clone>::clone(&x);
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
  |
help: box the return type, and wrap all of the returned values in `Box::new`
  |
3 |     Box<<::implicit_clone::ImplicitClone as Clone>::clone>(&x);
  |     ++++                                                 +

error[E0782]: trait objects must include the `dyn` keyword
 --> src/main.rs:3:6
  |
3 |     <::implicit_clone::ImplicitClone as Clone>::clone(&x);
  |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
help: add `dyn` keyword before this trait
  |
3 |     <dyn (::implicit_clone::ImplicitClone) as Clone>::clone(&x);
  |      +++++                               +

```

Reply:
`::implicit_clone::ImplicitClone::clone(&x)` doesn't work?
something like `Clone::clone(&x as &dyn ::implicit_clone::ImplicitClone)`?
`{ let _ = &x as &dyn ::implicit_clone::ImplicitClone; Clone::clone(&x) }`?
`{ fn implicit_clone<T: ImplicitClone>(value: &T) { value.clone() }; implicit_clone(&x) }` if nothing works? lol

Reply:
- `::implicit_clone::ImplicitClone::clone(&x)` no
- `Clone::clone(&x as &dyn ::implicit_clone::ImplicitClone)` no
- `{ let _: &dyn ::implicit_clone::ImplicitClone = &x; Clone::clone(&x) }` no
- `{ let _ = &x as &dyn ::implicit_clone::ImplicitClone; Clone::clone(&x) }` no
- `{ fn implicit_clone<T: ::implicit_clone::ImplicitClone>(value: &T) -> T { value.clone() } implicit_clone(&x) }` that worked!

Reply:
I feel it's not a very elegant solution :(

Reply:
Conceptually I'm not sure if this change is good actually. ImplicitClone is used by Yew to allow the type to be implicitly cloned when it's passed by reference in the `html!` macro. It's not used for "restricting" the input of the user. If I could have avoided introducing a new trait, I would have so I'm not sure that we actually *want* to restrict things here

Also the "help" message is replaced by an obscure "note":

```patch
diff --git a/tests/function_component_attr/autoprops-fail.stderr b/tests/function_component_attr/autoprops-fail.stderr
index d87685c..1aa9c95 100644
--- a/tests/function_component_attr/autoprops-fail.stderr
+++ b/tests/function_component_attr/autoprops-fail.stderr
@@ -41,19 +41,45 @@ note: struct `crate::private_module::CompPrivateTest` exists but is inaccessible
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not accessible
    = note: this error originates in the attribute macro `::yew::function_component` (in Nightly builds, run with -Z macro-backtrace for more info)
 
-error[E0277]: the trait bound `NotClonable: Clone` is not satisfied
+error[E0277]: the trait bound `NotClonable: ImplicitClone` is not satisfied
   --> tests/function_component_attr/autoprops-fail.rs:40:19
    |
 38 | #[autoprops]
    | ------------ required by a bound introduced by this call
 39 | #[function_component]
 40 | fn TypeIsNotClone(stuff: NotClonable) -> Html {
-   |                   ^^^^^ the trait `Clone` is not implemented for `NotClonable`
+   |                   ^^^^^ expected an implementor of trait `ImplicitClone`
    |
-help: consider annotating `NotClonable` with `#[derive(Clone)]`
+note: required by a bound in `<TypeIsNotClone as FunctionProvider>::run::inner::implicit_clone`
+  --> tests/function_component_attr/autoprops-fail.rs:38:1
    |
-36 | #[derive(Clone)]
+38 | #[autoprops]
+   | ^^^^^^^^^^^^ required by this bound in `<TypeIsNotClone as FunctionProvider>::run::inner::implicit_clone`
+   = note: this error originates in the attribute macro `autoprops` (in Nightly builds, run with -Z macro-backtrace for more info)
+help: consider borrowing here
+   |
+40 | fn TypeIsNotClone(&stuff: NotClonable) -> Html {
+   |                   +
+
+error[E0277]: the trait bound `NotPartialEq: ImplicitClone` is not satisfied
+  --> tests/function_component_attr/autoprops-fail.rs:52:23
+   |
+50 | #[autoprops]
+   | ------------ required by a bound introduced by this call
+51 | #[function_component]
+52 | fn TypeIsNotPartialEq(stuff: NotPartialEq) -> Html {
+   |                       ^^^^^ expected an implementor of trait `ImplicitClone`
+   |
+note: required by a bound in `<TypeIsNotPartialEq as FunctionProvider>::run::inner::implicit_clone`
+  --> tests/function_component_attr/autoprops-fail.rs:50:1
+   |
+50 | #[autoprops]
+   | ^^^^^^^^^^^^ required by this bound in `<TypeIsNotPartialEq as FunctionProvider>::run::inner::implicit_clone`
+   = note: this error originates in the attribute macro `autoprops` (in Nightly builds, run with -Z macro-backtrace for more info)
+help: consider borrowing here
    |
+52 | fn TypeIsNotPartialEq(&stuff: NotPartialEq) -> Html {
+   |                       +
 
 error[E0369]: binary operation `==` cannot be applied to type `NotPartialEq`
   --> tests/function_component_attr/autoprops-fail.rs:52:30
```

Reply:
The time was not wasted though because it did reveal another missing impl on primitive types: https://github.com/yewstack/implicit-clone/pull/42

Reply:
> I feel it's not a very elegant solution :(

That fn could be moved to ImplicitClone, like `std::mem::copy` and `std::mem::drop` (I thought `std::mem::move` also exists), all of which have a doc that says "This function is not magic; it is literally defined as..."

Reply:
Created https://github.com/yewstack/implicit-clone/pull/44

Reply:
Fixed in 45682cc